<a id="top"></a>
<h1 style="font-size:50px; color:#3A71AE"><center>Wrangle WeRateDogs</center></h1>

<h1 style="font-size:35px; color:#3A71AE">Table of Content:</h1>

1. [**Introduction**](#id_1)
2. [**Notebook Structure**](#id_2)
3. [**Data Wrangling Process**](#id_3)
4. [**Data Storing**](#id_4)

<a id="id_1"></a>
<h1 style="font-size:30px; color:#3A71AE">Introduction</h1>

[**Data wrangling**](https://en.wikipedia.org/wiki/Data_wrangling) or sometimes referred to as data munging, is the process of transforming and mapping data from one "raw" data form into another format with the intent of making it more appropriate and valuable for a variety of downstream purposes such as analytics.

**Note:** In this project there will be two notebooks, this one [**Wrangle WeRateDogs**](#top) that contains the `data wrangling` process and [**Analyze and Visualize WeRateDogs**](01_analyze_visualize.ipynb) that contains the visualization process.

Before anything, let's find out:
- [**The Dataset**](#id_1_1)
- [**Notebook Outline**](#id_1_2)
- [**Our Goal**](#id_1_3)


[Back to Top](#top)

<a id="id_1_1"></a>
## [The Dataset](#id_1_1)
**The dataset** we will be wrangling (and analyzing and visualizing) is the tweet archive of Twitter user [@dog_rates](https://twitter.com/dog_rates), also known as [WeRateDogs](https://en.wikipedia.org/wiki/WeRateDogs). WeRateDogs is a Twitter account that rates people's dogs with a humorous comment about the dog. These ratings almost always have a denominator of 10. The numerators, though? Almost always greater than 10 (`11/10`, `12/10`, `13/10`, etc).

[Back to Introduction](#id_1)

<a id="id_1_2"></a>
## [Notebook Outline](#id_1_2)
In this **Notebook** we will perform all of the processes below for [The Dataset](#id_1_1):
- [ ] Gathering
- [ ] Assessing
- [ ] Cleaning
- [ ] Storing
- [ ] Analyzing
- [ ] Visualizing
- [ ] Reporting<br><br>

[Back to Introduction](#id_1)

<a id="id_1_3"></a>
## [Our Goal](#id_1_3)
Our goal is to wrangle **WeRateDogs** Twitter data to create interesting and trustworthy analyses and visualizations.

[Back to Introduction](#id_1)

<a id="id_2"></a>
<h1 style="font-size:30px; color:#3A71AE">Notebook Structure</h1>

**This Notebook will be structured as follows:**<br>
- We will start with [Data Wrangling Process](#id_3) to [Gather](#id_3_1), [Assess](#id_3_2) and [Clean](#id_3_3) **WeRateDogs** tweets.
- After wrangling our data, we will storing it for any future purposes.
- Then the Analyzing and Visualizing will come.
- Finally, we will report our efforts.

[Back to Top](#top)

<a id="id_3"></a>
<h1 style="font-size:30px; color:#3A71AE">Data Wrangling Process</h1>

Data wrangling consists of three steps:
1. [**Gathering Data**](#id_3_1)
2. [**Assessing Data**](#id_3_2)
3. [**Cleaning Data**](#id_3_3)

[Back to Top](#top)

<a id="id_3_1"></a>
# [1. Gathering Data](#id_3_1)
[Gathering data](https://ori.hhs.gov/education/products/n_illinois_u/datamanagement/dctopic.html) is the first step in data wrangling. It is the process of collecting and measuring information on variables of interest, in an established systematic fashion that enables one to answer stated research questions, test hypotheses, and evaluate outcomes.

### Gathering data can be done in two ways:
- **Manually**, by simply click a button and download the file.
- **Programmatically**, through [web scraping](https://en.wikipedia.org/wiki/Web_scraping), [requesting URLs](https://requests.readthedocs.io/en/master/) or [APIs](https://www.mulesoft.com/resources/api/what-is-an-api).

### There are three types of data we need to collect for this projects:
- [**Enhanced Twitter Archive**](#id_3_1_1)
- [**Image Predictions File**](#id_3_1_2)
- [**Additional Data via the Twitter API**](#id_3_1_3)

[Back to Data Wrangling Process](#id_3)

<a id="id_3_1_1"></a>
## [Enhanced Twitter Archive](#id_3_1_1)
The [WeRateDogs](https://twitter.com/dog_rates?lang=ar) Twitter archive contains basic tweet data for Tweets with ratings only (there are 2356).

This data is ready, and all what we need to do is to **download** it **manually** through this link [twitter_archive_enhanced.csv](https://d17h27t6h515a5.cloudfront.net/topher/2017/August/59a4e958_twitter-archive-enhanced/twitter-archive-enhanced.csv) from [Udacity](udacity.com).

[Back to Gathering data](#id_3_1)

<a id="id_3_1_2"></a>
## [Image Predictions File](#id_3_1_2)
It's a file of predictions about what breed of dog (or other object, animal, etc.) is present in each tweet according to a neural network. 

This file `image_predictions.tsv` is hosted on Udacity's servers and we need to **download** it **programmatically** using the [Requests](https://pypi.org/project/requests/) library and the following URL:<br>
`https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv`

[Back to Gathering Data](#id_3_1)

In [66]:
# Import requests and os library to request the url and store it in data forder
import requests
import os

folder_data = 'data'

# Create folder named data, if it not exist, to store the file in
if not os.path.exists(folder_data):
    os.makedirs(folder_data)

# The url for "image_predictions.tsv" from Udacity
url = "https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv"

# request a response from Udacity server using request library
response = requests.get(url)

# displaying the respone contents
response.content[:500]

b'tweet_id\tjpg_url\timg_num\tp1\tp1_conf\tp1_dog\tp2\tp2_conf\tp2_dog\tp3\tp3_conf\tp3_dog\n666020888022790149\thttps://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg\t1\tWelsh_springer_spaniel\t0.465074\tTrue\tcollie\t0.156665\tTrue\tShetland_sheepdog\t0.0614285\tTrue\n666029285002620928\thttps://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg\t1\tredbone\t0.506826\tTrue\tminiature_pinscher\t0.07419169999999999\tTrue\tRhodesian_ridgeback\t0.07201\tTrue\n666033412701032449\thttps://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg\t1\tGerman_shepherd\t0.596461'

In [67]:
# Writing the response content to "image_predictions.tsv" in data folder
with open(os.path.join(folder_data, 'image_predictions.tsv'), 'wb') as f:
    f.write(response.content)

<a id="id_3_1_3"></a>
## [Additional Data via the Twitter API](#id_3_1_3)
We also need to gather, retweet count and favorite count. Fortunately, this additional data can be gathered by anyone from [Twitter's API](https://developer.twitter.com/en/docs). We're going to query Twitter's API to gather this valuable data.

**Note:** Twitter API data comes in `JSON` format.

[Back to Gathering Data](#id_3_1)

In [68]:
# Creating a function to display JSON files in a nice and human readable way
# Source: https://www.reddit.com/r/IPython/comments/34t4m7/lpt_print_json_in_collapsible_format_in_ipython/
import uuid
from IPython.display import display_javascript, display_html, display
import json

class RenderJSON(object):
    def __init__(self, json_data):
        if isinstance(json_data, dict):
            self.json_str = json.dumps(json_data)
        else:
            self.json_str = json_data
        self.uuid = str(uuid.uuid4())

    def _ipython_display_(self):
        display_html('<div id="{}" style="height: 600px; width:100%;"></div>'.format(self.uuid), raw=True)
        display_javascript("""
        require(["https://rawgit.com/caldwell/renderjson/master/renderjson.js"], function() {
        document.getElementById('%s').appendChild(renderjson(%s))
        });
        """ % (self.uuid, self.json_str), raw=True)

In [69]:
# Importing tweepy library to quary the wanted tweets from Twitter's APIsand pandas
import tweepy
import pandas as pd

# setting consumer key and secret, also, access token and secret for authentication purposes
consumer_key = "RQ71XK3zAUTenUz9M35OQlXcI"
consumer_secret = "za8Hby6CxIn6ROMVA4yYjKPfsK4BIeWuqNapAntoXkxgyCeYZ7"
access_token = "832729167867187201-Elnf7IEPzl7l6cmtF74w3p1iZ29xxTz"
access_secret = "rV0nrGqRiKKb4Ck3d24wmfZ3PqB0OdbkQCmrO4wiq73VH"

# Creating an OAuthHandler instance
auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

# Creating tweepy API class
api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

# Reading "twitter_archive_enhanced.csv" to get each tweet id
twitter_archive = pd.read_csv(os.path.join(folder_data, "twitter_archive_enhanced.csv"))

# Creating list for tweet ids
tweet_ids = list(twitter_archive['tweet_id'])

# Displaying the first tweet in twitter_archive df (as a JSON format) in a nice way using the predefined RenderJSON() function
RenderJSON(api.get_status(twitter_archive['tweet_id'][0], tweet_mode='extended')._json)

In [70]:
# Importing json library to read and write json files
import json

# The direction to save json file
tweet_json_dir = os.path.join(folder_data, 'tweet_json.txt')
# Delete the file if it exist in order to start fresh.
if os.path.exists(tweet_json_dir):
    os.remove(tweet_json_dir)

# Reading each tweet from tweets list and get its order number using enumerate()
# We need to maintain json file's structure as list of dictionaries, to load it later using json.load()
for id_number, tweet_id in enumerate(tweet_ids):
    try:
        # Getting each tweet status as json format
        tweet_json = api.get_status(tweet_id, tweet_mode='extended')._json
        
        # by using append mode, append each tweet's JSON data on its own line
        with open(tweet_json_dir, 'a') as file:
            # Adding ("[", json text, "," and new line) if it's the first json text
            if id_number == 0:
                file.write('[')
                json.dump(tweet_json, file)
                file.write(',\n')
                print(id_number, tweet_id)
                
            # Adding only (json text and "]") if it's the last json text
            elif id_number == (len(tweet_ids)-1):
                json.dump(tweet_json, file)
                file.write(']')
                print(id_number, tweet_id)
            
            # Adding (json text "," and new line) for anything between
            else:
                json.dump(tweet_json, file)
                file.write(',\n')
                print(id_number, tweet_id)
                
    # Catch errors and print it
    except Exception as e:
        print(id_number, e)

In [71]:
import pandas as pd
import numpy as np
from os.path import join
import json

# Reading "twitter_archive_enhanced" file as comma separated value file
enhanced_archive = pd.read_csv(join(folder_data, 'twitter_archive_enhanced.csv'))

# Reading "image_predictions" file as tab separated value file
image_predictions = pd.read_csv(join(folder_data, 'image_predictions.tsv'), sep='\t')

# Now let's read getretweet and favorite counts from "tweet_json.txt"
# 'tweets_json.txt' file
with open(join(folder_data, 'tweet_json.txt')) as file:
    tweets_json_txt = json.load(file)

# Create dataframe list
retweet_favorite_counts_ls = []

# for each tweet append tweet_id, retweet_count and favorite_count to dataframe_list
for tweet_json_txt in tweets_json_txt:
    retweet_favorite_counts_ls.append({'tweet_id' : tweet_json_txt['id'],
                                   'retweet_count' : tweet_json_txt['retweet_count'],
                                   'favorite_count' : tweet_json_txt['favorite_count']})

# Create a dataframe for retweet_count and favorite_count
retweet_favorite_counts = pd.DataFrame(retweet_favorite_counts_ls)

display(enhanced_archive.head(2))
display(image_predictions.head(2))
display(retweet_favorite_counts.head(2))

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                 text  \
0                                                       This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU   
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   

   retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
0                  NaN                       NaN                        NaN   
1                  NaN                       NaN                        NaN   

                                                     expanded_urls  \
0  https://twitter.com/dog_rates/status/892420643555336193/photo/1   
1  https://twitter.com/dog_rates/status/892177421306343426/photo/1   

   rating_numerator  rating_denominator     name doggo floofer pupper puppo  
0                13                  10  Phineas  None    None   None  None  
1                13                  10    Tilly  None    None   None  None

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True

tweet_id  retweet_count  favorite_count
0  892420643555336193           7565           35769
1  892177421306343426           5605           30898

<a id="id_3_1_4"></a>
> ### [So Far, So Good](#id_3_1_4)
Now we have `gathered` **manually** and **programmatically** all the data we wanted to reach [Our Goal](#id_1_3). So, let's proceed to [Assessing Data](#id_3_2) phase.
>- [x] Gathering
>- [ ] Assessing
>- [ ] Cleaning
>- [ ] Storing
>- [ ] Analyzing
>- [ ] Visualizing
>- [ ] Reporting
>
>[Back to Gathering Data](#id_3_1)

<a id="id_3_2"></a>
# [2. Assessing Data](#id_3_2)
Then let's start Assessing our data. We are going to inspect our dataset for two things, data quality issues (i.e. content issues) and lack of tidiness (i.e. structural issues), through two steps: detecting and documentation, by using two types of assessments: visual assessment and programmatic assessment.
 - [**Data Issues Types**](#id_3_2_1)
 - [**Assessment Steps**](#id_3_2_2)
 - [**Assessment Types**](#id_3_2_3)
 - [**Assessment Docummentions**](#id_3_2_4)
 
[Back to Data Wrangling Process](#id_3)

<a id="id_3_2_1"></a>
## [Data Issues Types](#id_3_2_1)
The two type of issues we are looking for are:
1. [**Quality Issues**](#id_3_2_1_1)
2. [**Structure Issues**](#id_3_2_1_2)
 
[Back to Assessing Data](#id_3_2)

<a id="id_3_2_1_1"></a>
### [1. Quality Issues](#id_3_2_1_1)
Data with issues in its content referred as `Dirty Data` or `Low Quality Data`.

**Common data quality issues include:**
- Missing data
- Invalid data
- Inaccurate data
- Inconsistent data

[Back to Data Issues Types](#id_3_2_1)

<a id="id_3_2_1_2"></a>
### [2. Structure Issues](#id_3_2_1_2)
Data that has issues with its structure is commonly referred as `Messy data` or ` Untidy Data`.

According to this [paper](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html), **In tidy data:**
- Each `variable` forms a `column`.
- Each `observation` forms a `row`.
- Each type of `observational unit` forms a `table`.

[Back to Data Issues Types](#id_3_2_1)

<a id="id_3_2_2"></a>
## [Assessment Steps](#id_3_2_2)
Assessing Phase has two steps:
1. **Detecting**, i.e. finding one of the previously defined [Data Issues](#id_3_2_1) in the dataset.
2. **Documenting**, i.e. documenting this issue as an `observation`.
 
[Back to Assessing Data](#id_3_2)

<a id="id_3_2_3"></a>
## [Assessment Types](#id_3_2_3)
There are two types of Data Assessing:
1. [**Visual Assessment**](#id_3_2_3_1)
2. [**Programmatic Assessment**](#id_3_2_3_2)
 
[Back to Assessing Data](#id_3_2)

<a id="id_3_2_3_1"></a>
### [1. Visual Assessment](#id_3_2_3_1)
Visual assessment is simple. Open the data in an application like (Google Sheets, Excel, a text editor,<br> Pandas `especially if the data is big`, etc.) and scroll through it, looking for [quality](#id_3_2_1_1) and [tidiness](#id_3_2_1_1) issues.<br><br>
<img src="https://assets-global.website-files.com/58e32bace1998d6e3fee8d74/5dfbe3b11bd3f56d754fcfa5_The-Google-Sheets-logo.-compressor.png" style="width:200px"></a><br><br>

The main purpose of **Visual Assessment** is to **acquaint** yourself with the dataset and just trying to understand it,<br> like **`acquiring a mental picture of it`**.
#### Visually Detecte Issues in:
- [Enhanced Twitter Archive](#id_3_2_3_1_1)
- [Image Predictions File](#id_3_2_3_1_2)
- [Additional Data via the Twitter API](#id_3_2_3_1_3)

**Note:** We are going to split our effort into two phases, first `detecting` data issues, then, `documenting` these issues in [Assessment Docummentions](#id_3_2_4) as defined in [Assessment Steps](#id_3_2_2).
 
[Back to Assessment Types](#id_3_2_3)

<a id="id_3_2_3_1_1"></a>
#### [Visually Detecte Issues in: Enhanced Twitter Archive](#id_3_2_3_1_1)

[Back to Visual Assessment](#id_3_2_3_1)<br>
[Advance to Assessment Docummentions](#id_3_2_4)

In [7]:
# Displaying enhanced_archive head and tail
enhanced_archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
0     2017-08-01 16:23:56 +0000   
1     2017-08-01 00:17:27 +0000   
2     2017-07-31 00:18:03 +0000   
3     2017-07-30 15:58:51 +0000   
4     2017-07-29 16:00:24 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
0     This is Phineas. He's a mystical boy. Only eve...                  NaN   
1     This is Tilly. She's just checking pup on you....                  NaN   
2     This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3     This is Darla. She commenced a snooze mid meal...                  NaN   
4     This is Franklin. He would like you to stop ca...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1     https://twitter.com/dog_rates/status/892177421...                13   
2     https://twitter.com/dog_rates/status/891815181...                12   
3     https://twitt

> **Notes:**
> - Variable `source` and (`<a href="`) string (anchor tag beginning) at the beginning of each value.
> - Variable `timestamp` data type is object not timedate.
> - Values `a` in variable `name`.
> - Values `None` in variable `name`.
> - Unnecessary "stages" variables (`doggo`, `floofer`, `pupper` and `puppo`)
> - "stages" variables values (`None`).
> 
> Now let's document these data issues in [Assessment Docummentions](#id_3_2_4)

<a id="id_3_2_3_1_2"></a>
#### [Visually Detecte Issues in: Image Predictions File](#id_3_2_3_1_2)
 
[Back to Visual Assessment](#id_3_2_3_1)<br>
[Advance to Assessment Docummentions](#id_3_2_4)

In [8]:
# Displaying image_predictions head and tail
image_predictions

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                      p1   p1_conf  p1_dog                  p2  \
0           1  Welsh_springer_spaniel  0.465074    True              collie   
1           1                 redbone  0.506826    True  miniature_pinscher   
2           1         German_shepherd  0.596461    True            malinois   
3           1     Rhodesian_ridgeback  0.408143    True             redbone   
4           1      miniature_pinscher  0.560311    True          Rottweiler   
...       ...                     ...       ...     ...                 ...   
2070        2                  basset  0.555712    True    English_springer   
2071        1             paper_towel  0.170278   False  Labrador_retriever   
2072        1               Chihuahua  0.716012    True            malamute   
2073        1               Chihuahua  0.323581    True            Pekinese   
2074        1                  orange  0.097049   False               bagel   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0     0.156665    True            Shetland_sheepdog  0.061428    True  
1     0.074192    True          Rhodesian_ridgeback  0.072010    True  
2     0.138584    True                   bloodhound  0.116197    True  
3     0.360687    True           miniature_pinscher  0.222752    True  
4     0.243682    True                     Doberman  0.154629    True  
...        ...     ...                          ...       ...     ...  
2070  0.225770    True  German_short-haired_pointer  0.175219    True  
2071  0.168086    True                      spatula  0.040836   False  
2072  0.078253    True                       kelpie  0.031379    True  
2073  0.090647    True                     papillon  0.068957    True  
2074  0.085851   False                       banana  0.076110   False  

[2075 rows x 12 columns]

> **Notes:**
>- Variables `p1`, `p2` and `p3` name values.
>- Variables `p1` through `p3_dog` are unnecessary.
>
>Now let's document these data issues in [Assessment Docummentions](#id_3_2_4)

<a id="id_3_2_3_1_3"></a>
#### [Visually Detecte Issues in: Additional Data via the Twitter API](#id_3_2_3_1_3)

[Back to Visual Assessment](#id_3_2_3_1)<br>
[Advance to Assessment Docummentions](#id_3_2_4)

In [9]:
# Displaying image_predictions head and tail
retweet_favorite_counts

tweet_id  retweet_count  favorite_count
0     892420643555336193           7565           35769
1     892177421306343426           5605           30898
2     891815181378084864           3715           23238
3     891689557279858688           7751           39065
4     891327558926688256           8346           37308
...                  ...            ...             ...
2326  666049248165822465             40              96
2327  666044226329800704            129             266
2328  666033412701032449             41             111
2329  666029285002620928             42             120
2330  666020888022790149            458            2383

[2331 rows x 3 columns]

> **Notes:**
>- This table is unnecessary.
>
>Now let's document these data issues in [Assessment Docummentions](#id_3_2_4)

<a id="id_3_2_3_2"></a>
### [2. Programmatic Assessment](#id_3_2_3_2)
**Programmatic assessment** is simply means using code to do anything other than looking to the dataset in its **`entirety`**. It tends to be more efficient than visual assessment. 

This types of assessments are handy for gauging data’s structure and also for quickly spotting things that we’ll need to clean.

**Note:** As We did in [Visually Assessment](#id_3_2_3_2), we'll start with `detecting` data issues, then, `documenting` these issues in [Assessment Docummentions](#id_3_2_4) as defined in [Assessment Steps](#id_3_2_2).

We are going to use [Pandas API](https://pandas.pydata.org/docs/reference/index.html) functions and methods in order to programmatically assess the dataset like:
- [DataFrame.sample](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sample.html?highlight=sample#pandas.DataFrame.sample)
- [Series.value_counts](https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html?highlight=value_counts#pandas.Series.value_counts)
- [DataFrame.info](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html?highlight=info#pandas.DataFrame.info)
- [DataFrame.dtypes](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dtypes.html?highlight=dtypes#pandas.DataFrame.dtypes)
- [DataFrame.describe](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html?highlight=describe#pandas.DataFrame.describe)
- [pandas.Series.isnull](https://pandas.pydata.org/docs/reference/api/pandas.Series.isnull.html?highlight=isnull#pandas.Series.isnull)
- [Series.notnull](https://pandas.pydata.org/docs/reference/api/pandas.Series.notnull.html?highlight=notnull#pandas.Series.notnull)

### Programmatically Detecte Issues in:
- [**Enhanced Twitter Archive**](#id_3_2_3_2_1)
- [**Image Predictions File**](#id_3_2_3_2_2)
- [**Additional Data via the Twitter API**](#id_3_2_3_2_3)

[Back to Assessment Types](#id_3_2_3)

<a id="id_3_2_3_2_1"></a>
#### [Programmatic Detecte Issues in: Enhanced Twitter Archive](#id_3_2_3_2_1)

[Back to Programmatic Assessment](#id_3_2_3_2)<br>
[Advance to Assessment Docummentions](#id_3_2_4)

#### Randomly checking the Data
[Back](#id_3_2_3_2_1)

In [10]:
# Display random samples from enhanced_archive
enhanced_archive.sample(20)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1419  698342080612007937                    NaN                  NaN   
994   748568946752774144                    NaN                  NaN   
2247  667873844930215936                    NaN                  NaN   
1914  674330906434379776           6.658147e+17           16374678.0   
1943  673709992831262724                    NaN                  NaN   
753   778774459159379968                    NaN                  NaN   
1635  684222868335505415                    NaN                  NaN   
751   779056095788752897                    NaN                  NaN   
474   816091915477250048                    NaN                  NaN   
200   854010172552949760                    NaN                  NaN   
1639  684177701129875456                    NaN                  NaN   
567   802239329049477120                    NaN                  NaN   
2153  669661792646373376                    NaN                  NaN   
932   754011816964026368                    NaN                  NaN   
2127  670319130621435904                    NaN                  NaN   
1013  747219827526344708                    NaN                  NaN   
1033  745074613265149952                    NaN                  NaN   
616   796484825502875648                    NaN                  NaN   
1861  675483430902214656                    NaN                  NaN   
1855  675531475945709568                    NaN                  NaN   

                      timestamp  \
1419  2016-02-13 03:05:01 +0000   
994   2016-06-30 17:28:39 +0000   
2247  2015-11-21 01:15:07 +0000   
1914  2015-12-08 20:53:11 +0000   
1943  2015-12-07 03:45:53 +0000   
753   2016-09-22 01:54:34 +0000   
1635  2016-01-05 04:00:18 +0000   
751   2016-09-22 20:33:42 +0000   
474   2017-01-03 01:20:49 +0000   
200   2017-04-17 16:34:26 +0000   
1639  2016-01-05 01:00:50 +0000   
567   2016-11-25 19:55:35 +0000   
2153  2015-11-25 23:39:47 +0000   
932   2016-07-15 17:56:40 +0000   
2127  2015-11-27 19:11:49 +0000   
1013  2016-06-27 00:07:44 +0000   
1033  2016-06-21 02:03:25 +0000   
616   2016-11-09 22:49:15 +0000   
1861  2015-12-12 01:12:54 +0000   
1855  2015-12-12 04:23:49 +0000   

                                                 source  \
1419  <a href="http://twitter.com/download/iphone" r...   
994   <a href="http://twitter.com/download/iphone" r...   
2247  <a href="http://twitter.com/download/iphone" r...   
1914  <a href="http://twitter.com/download/iphone" r...   
1943  <a href="http://twitter.com/download/iphone" r...   
753   <a href="http://twitter.com/download/iphone" r...   
1635  <a href="http://twitter.com/download/iphone" r...   
751   <a href="http://twitter.com/download/iphone" r...   
474   <a href="http://twitter.com/download/iphone" r...   
200   <a href="http://twitter.com/download/iphone" r...   
1639  <a href="http://twitter.com/download/iphone" r...   
567   <a href="http://twitter.com/download/iphone" r...   
2153  <a href="http://twitter.com/download/iphone" r...   
932   <a href="http://twitter.com/download/iphone" r...   
2127  <a href="http://twitter.com/download/iphone" r...   
1013  <a href="http://twitter.com/download/iphone" r...   
1033  <a href="http://vine.co" rel="nofollow">Vine -...   
616   <a href="http://twitter.com/download/iphone" r...   
1861  <a href="http://twitter.com/download/iphone" r...   
1855  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1419  This is Maximus. He's training for the tetherb...                  NaN   
994   This is Cora. She rings a bell for treats. 12/...                  NaN   
2247  Neat dog. Lots of spikes. Always in push-up po...                  NaN   
1914                                       13/10\n@ABC7                  NaN   
1943  I know a lot of you are studying for finals. G...                  NaN   
753   RT @dog_rates: In case you haven't seen t

#### Checking `source` and  `rating_denominator` variables.
[Back](#id_3_2_3_2_1)

In [11]:
# source values
enhanced_archive.source.value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       33
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

> After checking each of these URLs, I believe that variable came into our data by mistake due to a `human error`!!
    
[Advance to Assessment Docummentions](#id_3_2_4)

In [12]:
# Checking unique values in rating_denominator 
enhanced_archive.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

> Value 10 has the greater part of the variable. But there are other values too (50, 80, 110, etc.). Let's examine the url for some of these values to find out the reason of why the `rating_denominator` is something other than 10

In [13]:
# Examine values (11, 110 and 170)
print(enhanced_archive[enhanced_archive['rating_denominator'] == 11].expanded_urls.iloc[1].split(',')[0],
      enhanced_archive[enhanced_archive['rating_denominator'] == 110].expanded_urls.iloc[0],
      enhanced_archive[enhanced_archive['rating_denominator'] == 170].expanded_urls.iloc[0], sep= '\n')

https://twitter.com/dog_rates/status/740373189193256964/photo/1
https://twitter.com/dog_rates/status/684222868335505415/photo/1
https://twitter.com/dog_rates/status/731156023742988288/photo/1


> These values came either from when the program was fetching and gathering the dog rates as of this form `int` `/` `int`, but the form meant a date like the first URL (`9/11` attack), or the tweet date was before 2017, and there were no fixed standard then.<br><br>
So I think it would be better if we dropped it.

#### Checking variable `expanded_urls` values.
[Back](#id_3_2_3_2_1)

In [14]:
# checking expanded_urls var
enhanced_archive[enhanced_archive['rating_denominator'] == 11].expanded_urls.iloc[1]

'https://twitter.com/dog_rates/status/740373189193256964/photo/1,https://twitter.com/dog_rates/status/740373189193256964/photo/1,https://twitter.com/dog_rates/status/740373189193256964/photo/1,https://twitter.com/dog_rates/status/740373189193256964/photo/1'

> Repeated values in  `expanded_urls` . Let's [document this](#id_3_2_4).

#### Checking variable `name` values.
[Back](#id_3_2_3_2_1)

In [15]:
# Check values 'a' in name columns URLs
[url for url in enhanced_archive[enhanced_archive['name'] == 'a'].expanded_urls[:3]]

['https://twitter.com/dog_rates/status/881536004380872706/video/1',
 'https://twitter.com/dog_rates/status/792913359805018113/photo/1,https://twitter.com/dog_rates/status/792913359805018113/photo/1,https://twitter.com/dog_rates/status/792913359805018113/photo/1,https://twitter.com/dog_rates/status/792913359805018113/photo/1',
 'https://twitter.com/dog_rates/status/772581559778025472/photo/1,https://twitter.com/dog_rates/status/772581559778025472/photo/1,https://twitter.com/dog_rates/status/772581559778025472/photo/1']

> Appearently, the tweet with no dog names have `a` instead. Let's if there are any other name values fetched by mistake too.

In [16]:
# get only names with length less 3
short_names = enhanced_archive.name[enhanced_archive.name.map(lambda x: True if len(str(x)) < 3 else(False))].value_counts()
short_names

a     55
Bo     9
an     7
O      1
Mo     1
by     1
JD     1
Al     1
Jo     1
Ed     1
my     1
Name: name, dtype: int64

In [17]:
# examine some of these names URLs
for name in short_names.drop('a').index:
    print(name, enhanced_archive[enhanced_archive['name'] == name].expanded_urls.iloc[0])

Bo https://twitter.com/dog_rates/status/819004803107983360/photo/1,https://twitter.com/dog_rates/status/819004803107983360/photo/1,https://twitter.com/dog_rates/status/819004803107983360/photo/1,https://twitter.com/dog_rates/status/819004803107983360/photo/1,https://twitter.com/dog_rates/status/819004803107983360/photo/1,https://twitter.com/dog_rates/status/819004803107983360/photo/1,https://twitter.com/dog_rates/status/819004803107983360/photo/1,https://twitter.com/dog_rates/status/819004803107983360/photo/1
an https://twitter.com/dog_rates/status/703041949650034688/photo/1,https://twitter.com/dog_rates/status/703041949650034688/photo/1
O https://twitter.com/dog_rates/status/776201521193218049/photo/1
Mo https://twitter.com/dog_rates/status/825876512159186944/photo/1
by https://twitter.com/dog_rates/status/680085611152338944/photo/1,https://twitter.com/dog_rates/status/680085611152338944/photo/1,https://twitter.com/dog_rates/status/680085611152338944/photo/1
JD https://twitter.com/dog

> Well, some of these names (capital ones) are correct, they are just short names, but some of them need to be fixed as an example of **manually fixing quality issues** (like `JO` need to be `Jo. Jo` and `O` need to be `O'Malley`).
> With the other names, they are fetched mistakenly and need to be dropped (like `a`, `an`, `by`, etc.). 
> 
> Let's [document](#id_3_2_4) this and see if the names that have `a` values are there in the tweet text.

In [18]:
pd.set_option('display.max_colwidth', None) # Wide the columns to see the full values text

# Getting names that have values of 'a' and the text variable have 'named' string in it.
wrong_fetched_names = enhanced_archive[(enhanced_archive['name'] == 'a') & 
                                       (enhanced_archive['text'].str.contains("named"))][['text', 'name']]
wrong_fetched_names

text  \
1853  This is a Sizzlin Menorah spaniel from Brooklyn named Wylie. Lovable eyes. Chiller as hell. 10/10 and I'm out.. poof https://t.co/7E0AiJXPmI   
1955   This is a Lofted Aphrodisiac Terrier named Kip. Big fan of bed n breakfasts. Fits perfectly. 10/10 would pet firmly https://t.co/gKlLpNzIl3   
2034      This is a Tuscaloosa Alcatraz named Jacob (Yacōb). Loves to sit in swing. Stellar tongue. 11/10 look at his feet https://t.co/2IslQ8ZSc7   
2066  This is a Helvetica Listerine named Rufus. This time Rufus will be ready for the UPS guy. He'll never expect it 9/10 https://t.co/34OhVhMkVr   
2116   This is a Deciduous Trimester mix named Spork. Only 1 ear works. No seat belt. Incredibly reckless. 9/10 still cute https://t.co/CtuJoLHiDo   
2125     This is a Rich Mahogany Seltzer named Cherokee. Just got destroyed by a snowball. Isn't very happy about it. 9/10 https://t.co/98ZBi6o4dj   
2128    This is a Speckled Cauliflower Yosemite named Hemry. He's terrified of intruder dog. Not one bit comfortable. 9/10 https://t.co/yV3Qgjh8iN   
2146    This is a spotted Lipitor Rumpelstiltskin named Alphred. He can't wait for the Turkey. 10/10 would pet really well https://t.co/6GUGO7azNX   
2161    This is a Coriander Baton Rouge named Alfredo. Loves to cuddle with smaller well-dressed dog. 10/10 would hug lots https://t.co/eCRdwouKCl   
2191  This is a Slovakian Helter Skelter Feta named Leroi. Likes to skip on roofs. Good traction. Much balance. 10/10 wow! https://t.co/Dmy2mY2Qj5   
2218    This is a Birmingham Quagmire named Chuk. Loves to relax and watch the game while sippin on that iced mocha. 10/10 https://t.co/HvNg9JWxFt   
2235             This is a Trans Siberian Kellogg named Alfonso. Huge ass eyeballs. Actually Dobby from Harry Potter. 7/10 https://t.co/XpseHBlAAb   
2249  This is a Shotokon Macadamia mix named Cheryl. Sophisticated af. Looks like a disappointed librarian. Shh (lol) 9/10 https://t.co/J4GnJ5Swba   
2255      This is a rare Hungarian Pinot named Jessiga. She is either mid-stroke or got stuck in the washing machine. 8/10 https://t.co/ZU0i0KJyqD   
2264                        This is a southwest Coriander named Klint. Hat looks expensive. Still on house arrest :(\n9/10 https://t.co/IQTOMqDUIe   
2273    This is a northern Wahoo named Kohl. He runs this town. Chases tumbleweeds. Draws gun wicked fast. 11/10 legendary https://t.co/J4vn2rOYFk   
2304                      This is a curly Ticonderoga named Pepe. No feet. Loves to jet ski. 11/10 would hug until forever https://t.co/cyDfaK8NBc   
2311                                   This is a purebred Bacardi named Octaviath. Can shoot spaghetti out of mouth. 10/10 https://t.co/uEvsGLOFHa   
2314   This is a golden Buckminsterfullerene named Johm. Drives trucks. Lumberjack (?). Enjoys wall. 8/10 would hug softly https://t.co/uQbZJM2DQB   

     name  
1853    a  
1955    a  
2034    a  
2066    a  
2116    a  
2125    a  
2128    a  
2146    a  
2161    a  
2191    a  
2218    a  
2235    a  
2249    a  
2255    a  
2264    a  
2273    a  
2304    a  
2311    a  
2314    a

> Apparently there are some name that has wrongly fetched need to be corrected. Let's [document](#id_3_2_4) this issue, and see if there are any incorect names.

In [19]:
# Getting unique names
enhanced__unique_names = enhanced_archive.name.unique()
enhanced__unique_names

array(['Phineas', 'Tilly', 'Archie', 'Darla', 'Franklin', 'None', 'Jax',
       'Zoey', 'Cassie', 'Koda', 'Bruno', 'Ted', 'Stuart', 'Oliver',
       'Jim', 'Zeke', 'Ralphus', 'Canela', 'Gerald', 'Jeffrey', 'such',
       'Maya', 'Mingus', 'Derek', 'Roscoe', 'Waffles', 'Jimbo', 'Maisey',
       'Lilly', 'Earl', 'Lola', 'Kevin', 'Yogi', 'Noah', 'Bella',
       'Grizzwald', 'Rusty', 'Gus', 'Stanley', 'Alfy', 'Koko', 'Rey',
       'Gary', 'a', 'Elliot', 'Louis', 'Jesse', 'Romeo', 'Bailey',
       'Duddles', 'Jack', 'Emmy', 'Steven', 'Beau', 'Snoopy', 'Shadow',
       'Terrance', 'Aja', 'Penny', 'Dante', 'Nelly', 'Ginger', 'Benedict',
       'Venti', 'Goose', 'Nugget', 'Cash', 'Coco', 'Jed', 'Sebastian',
       'Walter', 'Sierra', 'Monkey', 'Harry', 'Kody', 'Lassie', 'Rover',
       'Napolean', 'Dawn', 'Boomer', 'Cody', 'Rumble', 'Clifford',
       'quite', 'Dewey', 'Scout', 'Gizmo', 'Cooper', 'Harold', 'Shikha',
       'Jamesy', 'Lili', 'Sammy', 'Meatball', 'Paisley', 'Albus',
       'Nept

> There are some other incorrect names, e.g., `this`, `his`, `old`, etc. And they all are `lowercase` not uppercase, so, let's check lowercase names.

In [20]:
# Getting only lowe case names 
wrong_names = enhanced_archive[enhanced_archive.name.str.islower()]
wrong_names = wrong_names['name'].unique()
wrong_names

array(['such', 'a', 'quite', 'not', 'one', 'incredibly', 'mad', 'an',
       'very', 'just', 'my', 'his', 'actually', 'getting', 'this',
       'unacceptable', 'all', 'old', 'infuriating', 'the', 'by',
       'officially', 'life', 'light', 'space'], dtype=object)

> Let's [document](#id_3_2_4) this quality issue too.

#### Checking multiple values in `stage` variables.
[Back](#id_3_2_3_2_1)

In [21]:
# Let's check if there any dog with multiple values in stages variables
# tweet_id var + stage vars
repeated_stage_cols = ['tweet_id', 'doggo', 'floofer', 'pupper', 'puppo']

# Coy enhanced_archive_clean for only the previous cols
repeated_stage = enhanced_archive.copy()[repeated_stage_cols]

# change stage col values to 0 if None and 1 if else
for col in repeated_stage_cols[1:]:
    repeated_stage[col] = repeated_stage[col].apply(lambda x: 0 if x == 'None' else(1))

# sum stage cols values to see if there any values greater than 1
repeated_stage['repeated'] = repeated_stage[repeated_stage_cols[1:]].apply(np.sum, axis=1)

# check value counts
repeated_stage.repeated.value_counts()

0    1976
1     366
2      14
Name: repeated, dtype: int64

> Yep, it looks like we have 14 dogs that multiple values for stage. Le'ts check some of these.

In [22]:
repeated_stage.query('repeated > 1')

tweet_id  doggo  floofer  pupper  puppo  repeated
191   855851453814013952      1        0       0      1         2
200   854010172552949760      1        1       0      0         2
460   817777686764523521      1        0       1      0         2
531   808106460588765185      1        0       1      0         2
565   802265048156610565      1        0       1      0         2
575   801115127852503040      1        0       1      0         2
705   785639753186217984      1        0       1      0         2
733   781308096455073793      1        0       1      0         2
778   775898661951791106      1        0       1      0         2
822   770093767776997377      1        0       1      0         2
889   759793422261743616      1        0       1      0         2
956   751583847268179968      1        0       1      0         2
1063  741067306818797568      1        0       1      0         2
1113  733109485275860992      1        0       1      0         2

In [23]:
print(enhanced_archive.query('tweet_id == "855851453814013952"').expanded_urls.iloc[0],
      enhanced_archive.query('tweet_id == "817777686764523521"').expanded_urls.iloc[0],
      enhanced_archive.query('tweet_id == "733109485275860992"').expanded_urls.iloc[0], sep='\n')

https://twitter.com/dog_rates/status/855851453814013952/photo/1
https://twitter.com/dog_rates/status/817777686764523521/video/1
https://twitter.com/dog_rates/status/733109485275860992/photo/1


> This's another `human error` during **Gathering** Enhanced Archive data, because sometime the same tweet could have multiple values that considered stage values, e.g. `puppo` and `doggo` in the first tweet. I'm going to use stage as part of my future analysis, so it would be better dropping these records. Let's [document this](#id_3_2_4).

#### Checking missing values in `enhanced_archive` table.
[Back](#id_3_2_3_2_1)

In [24]:
# Checking missing values
enhanced_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

> The missing values are due to the variables nature, i.e. if the tweet was in reply to status or user, and if it got a retweet. <br>
Except for `expanded_urls` variable missing values, It could be because of `human errros`. Let's check those null values using `isnull()` method and `boolean indexing`.

#### Checking missing values in `expanded_urls` variable.
[Back](#id_3_2_3_2_1)

In [25]:
enhanced_archive[enhanced_archive['expanded_urls'].isnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
30    886267009285017600           8.862664e+17         2.281182e+09   
55    881633300179243008           8.816070e+17         4.738443e+07   
64    879674319642796034           8.795538e+17         3.105441e+09   
113   870726314365509632           8.707262e+17         1.648776e+07   
148   863427515083354112           8.634256e+17         7.759620e+07   
179   857214891891077121           8.571567e+17         1.806710e+08   
185   856330835276025856                    NaN                  NaN   
186   856288084350160898           8.562860e+17         2.792810e+08   
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
218   850333567704068097           8.503288e+17         2.195506e+07   
228   848213670039564288           8.482121e+17         4.196984e+09   
234   847617282490613760           8.476062e+17         4.196984e+09   
274   840698636975636481           8.406983e+17         8.405479e+17   
290   838150277551247360           8.381455e+17         2.195506e+07   
291   838085839343206401           8.380855e+17         2.894131e+09   
313   835246439529840640           8.352460e+17         2.625958e+07   
342   832088576586297345           8.320875e+17         3.058208e+07   
346   831926988323639298           8.319030e+17         2.068372e+07   
375   828361771580813312                    NaN                  NaN   
387   826598799820865537           8.265984e+17         4.196984e+09   
409   823333489516937216           8.233264e+17         1.582854e+09   
427   821153421864615936           8.211526e+17         1.132119e+08   
498   813130366689148928           8.131273e+17         4.196984e+09   
513   811647686436880384           8.116272e+17         4.196984e+09   
570   801854953262350336           8.018543e+17         1.185634e+07   
576   800859414831898624           8.008580e+17         2.918590e+08   
611   797165961484890113           7.971238e+17         2.916630e+07   
701   786051337297522688           7.727430e+17         7.305050e+17   
707   785515384317313025                    NaN                  NaN   
843   766714921925144576           7.667118e+17         4.196984e+09   
857   763956972077010945           7.638652e+17         1.584641e+07   
967   750381685133418496           7.501805e+17         4.717297e+09   
1005  747651430853525504           7.476487e+17         4.196984e+09   
1080  738891149612572673           7.384119e+17         3.589728e+08   
1295  707983188426153984           7.079801e+17         2.319108e+09   
1345  704491224099647488           7.044857e+17         2.878549e+07   
1445  696518437233913856                    NaN                  NaN   
1446  696490539101908992           6.964887e+17         4.196984e+09   
1474  693644216740769793           6.936422e+17         4.196984e+09   
1479  693582294167244802           6.935722e+17         1.198989e+09   
1497  692423280028966913           6.924173e+17         4.196984e+09   
1523  690607260360429569           6.903413e+17         4.670367e+08   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1605  685681090388975616           6.855479e+17         4.196984e+09   
1618  684969860808454144           6.849598e+17         4.196984e+09   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1689  681340665377193984           6.813394e+17         4.196984e+09   
1774  678023323247357953           6.780211e+17         4.196984e+09   
1819  676590572941893632           6.765883e+17         4.196984e+09   
1844  675849018447167488           6.758457e+17         4.196984e+09   
1895  674742531037511680           6.747400e+17         4.196984e+09   
1905  674606911342424069           6.744689e+17         4.196984e+09   
1914  674330906434379776           6.658147e+17         1.637468e+07   
1940  673716320723169284           6.737159e+17         

> We can examine one of these tweets original **JSON** text using tweet_id and `tweets_json_txt` we created early. Then, render it using the predifined function `RenderJSON`.

In [26]:
# Get an id for these missing expanded_urls tweets
missing_id = enhanced_archive[enhanced_archive['expanded_urls'].isnull()]['tweet_id'].iloc[0]

# Let's get one of these tweets JOSN text
for json_object in tweets_json_txt:
    if json_object['id'] == missing_id:
        mssing_json = json_object
        print('Found Successfully')

# Render the tweet JSON text
RenderJSON(mssing_json)

Found Successfully


> It's looks like the `expanded_urls` is missing too in JSON text, thus, there's nothing we can do here.

#### Checking  table `enhanced_archive` datatypes.
[Back](#id_3_2_3_2_1)

In [27]:
enhanced_archive.dtypes

tweet_id                        int64
in_reply_to_status_id         float64
in_reply_to_user_id           float64
timestamp                      object
source                         object
text                           object
retweeted_status_id           float64
retweeted_status_user_id      float64
retweeted_status_timestamp     object
expanded_urls                  object
rating_numerator                int64
rating_denominator              int64
name                           object
doggo                          object
floofer                        object
pupper                         object
puppo                          object
dtype: object

> **Note: data types for these variables:**<br>
>    - tweet_id
>    - in_reply_to_status_id
>    - in_reply_to_user_id
>    - retweeted_status_id
>    - retweeted_status_user_id
>    - retweeted_status_timestamp
>
>[Advance to Assessment Docummentions](#id_3_2_4)

<a id="id_3_2_3_2_2"></a>
#### [Programmatic Detecte Issues in: Image Predictions File](#id_3_2_3_2_2)
 
[Back to Programmatic Assessment](#id_3_2_3_2)<br>
[Advance to Assessment Docummentions](#id_3_2_4)

#### Randomly checking the Data in `image_predictions` table
[Back](#id_3_2_3_2_2)

In [28]:
# Display random samples from image_predictions
image_predictions.sample(20)

tweet_id  \
946   704499785726889984   
1654  810254108431155201   
966   706291001778950144   
140   668544745690562560   
1260  748705597323898880   
70    667192066997374976   
1552  793150605191548928   
133   668480044826800133   
1647  808733504066486276   
1217  743595368194129920   
1320  756288534030475264   
1606  800459316964663297   
2044  886258384151887873   
2014  879415818425184262   
660   682406705142087680   
1056  714606013974974464   
36    666447344410484738   
575   678643457146150913   
1199  740711788199743490   
426   674082852460433408   

                                                                                      jpg_url  \
946                                           https://pbs.twimg.com/media/Ccbi0UGWoAA4fwg.jpg   
1654                                          https://pbs.twimg.com/media/Cz6Z0DgWIAAfdvp.jpg   
966                                           https://pbs.twimg.com/media/Cc0_2tXXEAA2iTY.jpg   
140                                           https://pbs.twimg.com/media/CUcl5jeWsAA6ufS.jpg   
1260  https://pbs.twimg.com/ext_tw_video_thumb/748704826305970176/pu/img/QHuadM5eEygfBeOf.jpg   
70                                            https://pbs.twimg.com/media/CUJXpRBXIAAN0yz.jpg   
1552                                          https://pbs.twimg.com/media/CwHWOZ7W8AAHv8S.jpg   
133                                           https://pbs.twimg.com/media/CUbrDWOWcAEyMdM.jpg   
1647                                          https://pbs.twimg.com/media/Czky0v9VIAEXRkd.jpg   
1217                                          https://pbs.twimg.com/media/ClHICHmXEAI_1PS.jpg   
1320                                          https://pbs.twimg.com/media/Cn7gaHrWIAAZJMt.jpg   
1606                                          https://pbs.twimg.com/media/CxvNfrhWQAA2hKM.jpg   
2044                                          https://pbs.twimg.com/media/DEyfTG4UMAE4aE9.jpg   
2014  https://pbs.twimg.com/ext_tw_video_thumb/879415784908390401/pu/img/cX7XI1TnUsseGET5.jpg   
660                                           https://pbs.twimg.com/media/CXhlRmRUMAIYoFO.jpg   
1056                                          https://pbs.twimg.com/media/CerKYG8WAAM1aE-.jpg   
36                                            https://pbs.twimg.com/media/CT-yU5QWwAEjLX5.jpg   
575                                           https://pbs.twimg.com/media/CWsGnyMVEAAM1Y1.jpg   
1199                                          https://pbs.twimg.com/media/CkeJcNkXEAAcrks.jpg   
426                                           https://pbs.twimg.com/media/CVrSxy7WsAAFD2F.jpg   

      img_num                         p1   p1_conf  p1_dog  \
946         1                  Chihuahua  0.376541    True   
1654        1  Staffordshire_bullterrier  0.292556    True   
966         1             Border_terrier  0.587101    True   
140         1                   bearskin  0.427870   False   
1260        1                tiger_shark  0.548497   False   
70          1                 Rottweiler  0.283640    True   
1552        1          Italian_greyhound  0.193869    True   
133         1                 Arctic_fox  0.119243   False   
1647        1                  seat_belt  0.779137   False   
1217        1               hippopotamus  0.505675   False   
1320        3                      conch  0.925621   False   
1606        1                      teddy  0.311928   False   
2044        1                        pug  0.943575    True   
2014        1           English_springer  0.383404    True   
660         1                     wombat  0.709344   False   
1056        1            Norfolk_terrier  0.293007    True   
36          1     curly-coated_retriever  0.322084    True   
575         1         Labrador_retriever  0.338757    True   
1199        1                 toy_poodle  0.388277    True   
426         1                 Pomeranian  0.666957    True   

                                  p2   p2_conf  p2_dog  \
946                      Sia

#### Checking missing values in `image_predictions` table
[Back](#id_3_2_3_2_2)

In [29]:
# Checking missing values
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


#### Checking variable `p1` values.
[Back](#id_3_2_3_2_2)

In [30]:
# Check p1 values
image_predictions.p1.value_counts()

golden_retriever      150
Labrador_retriever    100
Pembroke               89
Chihuahua              83
pug                    57
                     ... 
grille                  1
dhole                   1
hare                    1
quilt                   1
hay                     1
Name: p1, Length: 378, dtype: int64

#### Checking table `image_predictions` data types.
[Back](#id_3_2_3_2_2)

In [31]:
# Check data types
image_predictions.dtypes

tweet_id      int64
jpg_url      object
img_num       int64
p1           object
p1_conf     float64
p1_dog         bool
p2           object
p2_conf     float64
p2_dog         bool
p3           object
p3_conf     float64
p3_dog         bool
dtype: object

> **Note:**<br>
>- tweet_id data type
>    
>Now let's document these data issues in [Assessment Docummentions](#id_3_2_4)

<a id="id_3_2_3_2_3"></a>
#### [Programmatic Detecte Issues in: Additional Data via the Twitter API](#id_3_2_3_2_3)

[Back to Programmatic Assessment](#id_3_2_3_2)<br>
[Advance to Assessment Docummentions](#id_3_2_4)

#### Randomly checking the Data in `retweet_favorite_counts` table.
[Back](#id_3_2_3_2_3)

In [32]:
# Display random samples from retweet_favorite_counts
retweet_favorite_counts.sample(20)

tweet_id  retweet_count  favorite_count
563   800443802682937345           4353               0
889   757596066325864448           1046            4308
23    887343217045368832           9411           31230
849   761599872357261312           1182            4108
360   828361771580813312            173            2188
1430  695446424020918272           1776            4330
1143  721503162398597120           1787            4526
1481  691793053716221953           4141            8005
1577  686003207160610816            612            1797
1304  705970349788291072            878            3117
1015  744234799360020481          76445          154179
866   759846353224826880           1949            6699
688   785533386513321988           2010            9179
726   780192070812196864           2241            8728
671   787717603741622272           2789           10336
1168  717790033953034240           1085            2824
1156  719339463458033665           1204            4346
999   746521445350707200            971               0
2122  669753178989142016            372             759
1920  673707060090052608            371            1048

#### Checking the missing values in `retweet_favorite_counts` table.
[Back](#id_3_2_3_2_3)

In [33]:
# Checking missing values
retweet_favorite_counts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   tweet_id        2331 non-null   int64
 1   retweet_count   2331 non-null   int64
 2   favorite_count  2331 non-null   int64
dtypes: int64(3)
memory usage: 54.8 KB


#### Statistically describe `retweet_favorite_counts` table.
[Back](#id_3_2_3_2_3)

In [34]:
# Check table describe stats
retweet_favorite_counts.describe()

tweet_id  retweet_count  favorite_count
count  2.331000e+03    2331.000000     2331.000000
mean   7.419079e+17    2660.208494     7471.366795
std    6.823170e+16    4498.541837    11601.403978
min    6.660209e+17       1.000000        0.000000
25%    6.782670e+17     539.000000     1300.500000
50%    7.182469e+17    1244.000000     3247.000000
75%    7.986692e+17    3093.500000     9150.500000
max    8.924206e+17   76445.000000   154179.000000

#### Checking table `retweet_favorite_counts` data types.
[Back](#id_3_2_3_2_3)

In [35]:
# Check data types
retweet_favorite_counts.dtypes

tweet_id          int64
retweet_count     int64
favorite_count    int64
dtype: object

> `tweet_id` variable data type should be object but we are going to merge the table with `enhanced_archive` table anyway.
>
>[Back to Programmatic Assessment](#id_3_2_3_2)<br>
>[Advance to Assessment Docummentions](#id_3_2_4)

<a id="id_3_2_4"></a>
## [Assessment Docummentions](#id_3_2_4)

Let's seperate our docummentions by data issues types, [Quality Issues](#id_3_2_4_1) and [Tidiness Issues](#id_3_2_4_2), then group it by each dataset:

<a id="id_3_2_4_1"></a>
### [Quality Issues:](#id_3_2_4_1)
1. #### Enhanced Twitter Archive - [Cleaning Phase](#id_3_3_3_3_1)
[Visual Assess](#id_3_2_3_1_1) - [Programmatic Assess](#id_3_2_3_2_1)
    - Variables `source` and `rating_denominator` are useless in our data.
    - There are some names with values of `a`, some of them need to be dropped and the other (`wrong_fetched_names` dataframe) need to be modified from the `text` column.
    - Some of name values need to be corrected like `JO` need to be `Jo. Jo` and `O` need to be `O'Malley`.
    - Some of name values need to be dropped like `such`, `a`, `quite`, `not`, `one`, `incredibly` and `mad`, etc.
    - Values `None` in variable `name` (indices 2351 and 2355).
    - Stage variables values, string `None`.
    - The new created variable `stage` need to be categorical data type.
    - Some records in stage variables have cross duplication (1 record with multiple stage values).
    - Repeated values for the same extended_urls records.
    - data types for these variables need to be change to `Object` data type except for `retweeted_status_timestamp` and `timestamp` which have to be in datetime format:
        - tweet_id
        - in_reply_to_status_id
        - in_reply_to_user_id
        - retweeted_status_id
        - retweeted_status_user_id
        - timestamp
        - retweeted_status_timestamp
    - After merging `enhanced_archive_clean` and `retweet_favorite_counts_clean`, there are some missing values in `retweet_counts` and `favorite_counts` columns.
        
- #### Image Predictions File  - [Cleaning Phase](#id_3_3_3_3_2)
[Visual Assess](#id_3_2_3_1_2) - [Programmatic Assess](#id_3_2_3_2_2)
    - Inconsistent name values in `p1`, `p2` and `p3` variables some with character "_" and some with lowercase.
    - The new created variable after melting, `number` as the number of prediction need to be cateforical data type.
    - Variable `tweet_id` data type need to be changed from `int64` to `object`.

<a id="id_3_2_4_2"></a>
### [Tidiness Issues:](#id_3_2_4_2)
1. #### Enhanced Twitter Archive - [Visual Assess](#id_3_2_3_1_1) - [Cleaning Phase](#id_3_3_3_2_2)
    - The 4 dog "stages" variables (`doggo`, `floofer`, `pupper` and `puppo`) are unnecessary, and can be reduced to one variable named `stage`.
    
- #### Image Predictions File - [Visual Assess](#id_3_2_3_1_2) - [Cleaning Phase](#id_3_3_3_2_3)
    - Variables `p1` through `p3_dog` are unnecessary, and can be reduced to 4 variables (`prediction_num`, `prediction_value`, `prediction_conf` and `prediction_reality`).

- #### Additional Data via the Twitter API - [Visual Assess](#id_3_2_3_1_3) - [Cleaning Phase](#id_3_3_3_2_1)
    - This table is unnecessary and can be included in the [Enhanced Twitter Archive](#id_3_2_3_1_1) table.

[Back to Assessing Data](#id_3_2)<br>
[Advance to Data Cleaning Process](#id_3_3_3)

<a id="id_3_2_5"></a>
> ### [So Far, So Good](#id_3_2_5)
>After we have `assessed` the dataset **manually** and **programmatically**, we're ready for the final step in Data Wrangling, [Cleaning Data](#id_3_3).
>- [X] Gathering
>- [X] Assessing
>- [ ] Cleaning
>- [ ] Storing
>- [ ] Analyzing
>- [ ] Visualizing
>- [ ] Reporting
>
>[Back to Data Wrangling Process](#id_3)

<a id="id_3_3"></a>
# [3. Cleaning Data](#id_3_3)
[Cleaning Data](https://en.wikipedia.org/wiki/Data_cleansing) is the final step in Data Wrangling. It's the process of correcting (or removing) corrupt or inaccurate records from a record set, table, or database through replacing, modifying, or deleting the dirty.

- [**Types of Data Cleaning**](#id_3_3_1)
- [**Data Cleaning Steps**](#id_3_3_2)
- [**Data Copying**](#id_3_3_copy)
- [**Data Cleaning Process**](#id_3_3_3)

<a id="id_3_3_1"></a>
## [Two types of Data Cleaning:](#id_3_3_1)
- **Manually** Cleaning (inefficient, time consuming and error prone)
- **Programmatically** Cleaning (for issues that occur more than one-time )

<a id="id_3_3_2"></a>
## [Data Cleaning Steps:](#id_3_3_2)
- **Define**, it's using verbs in order to call an action to fix the issue we observed.
- **Code**, it's doing the action we specified using programming codes.
- **Test**, if our code worked, using [visually](#id_3_2_3_1) or [programmatically](#id_3_2_3_2) assessments.

[Back to Data Wrangling](#id_3)

<a id="id_3_3_copy"></a>
## [Data Copying](#id_3_3_copy)
**Note:** Do not forget to copy each table before starting **Cleaning** phase, so, we can easily back to the original table of need.

Copying each dataframe using [DataFrame.copy](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.copy.html)

[Back to Cleaning Data](#id_3_3)

<a id="id_3_3_3"></a>
# [Data Cleaning Process:](#id_3_3_3)
We have 2 data issues need to be cleaned before continuing Analyzing and Visualizing
1. [Missing Values](#id_3_3_3_1)
2. [Cleaning for Tidiness](#id_3_3_3_2)
3. [Cleaning for Quality](#id_3_3_3_3)

**Note:** Don't forget to [**copy**](#id_3_3_copy) data before statring, so, we can back to the original data if we want.

[Back to Data Wrangling Process](#id_3)<br>
[Back to Assessment Docummentions](#id_3_2_4)

In [36]:
# Copying each dataframe
enhanced_archive_clean = enhanced_archive.copy()
image_predictions_clean = image_predictions.copy()
retweet_favorite_counts_clean = retweet_favorite_counts.copy()

<a id="id_3_3_3_1"></a>
# [1. Missing Values](#id_3_3_3_1)
Let's start with missing values in enhanced_archive_clean dataframe.

[Define](#id_3_3_3_1_1) [Code](#id_3_3_3_1_2) [Test](#id_3_3_3_1_3)

[Back to Data Cleaning Process](#id_3_3_3)<br>
[Back to Assessment Docummentions](#id_3_2_4)

<a id="id_3_3_3_1_1"></a>
#### [Define](#id_3_3_3_1_1)
Replace all the `None` values in enhanced_archive_clean data frame with [numpy.nan](https://numpy.org/doc/stable/reference/constants.html?highlight=nan#numpy.NaN) values.

[Back](#id_3_3_3_1)

<a id="id_3_3_3_1_2"></a>
#### [Code](#id_3_3_3_1_2)
[Back](#id_3_3_3_1)

In [37]:
# Replaceing Nones with NaNs
import numpy as np

# Replaceing None with NaNs
enhanced_archive_clean.replace('None', np.nan, inplace=True)

<a id="id_3_3_3_1_3"></a>
#### [Test](#id_3_3_3_1_3)
[Back](#id_3_3_3_1)

In [38]:
enhanced_archive_clean.info() # Check variable data types

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        1611 non-null   object 
 13  doggo                       97 no

<a id="id_3_3_3_2"></a>
# [2. Cleaning for Tidiness](#id_3_3_3_2)
Let's start cleaning the dataset and we going to start with structure issues"
1. [Merging Enhanced Archive DataFrame with Retweet and favorite DataFrame](#id_3_3_3_2_1)
2. [Metling Dog Stage Columns](#id_3_3_3_2_2)
3. [Melting Predictions Columns](#id_3_3_3_2_3)


[Back to Data Cleaning Process](#id_3_3_3)<br>
[Back to Assessment Docummentions](#id_3_2_4)

<a id="id_3_3_3_2_1"></a>
### [Merging Enhanced Archive DataFrame with Retweet and favorite DataFrame](#id_3_3_3_2_1)
[Define](#id_3_3_3_2_1_1) [Code](#id_3_3_3_2_1_2) [Test](#id_3_3_3_2_1_3)

[Back to Cleaning for Tidiness](#id_3_3_3_2)<br>
[Back to Docummentions](#id_3_2_4_2)

<a id="id_3_3_3_2_1_1"></a>
#### [Define](#id_3_3_3_2_1_1)
Merge `enhanced_archive_clean` and `retweet_favorite_counts_clean` on `tweet_id` using [DataFrame.merge]() method.

[Back](#id_3_3_3_2_1)

<a id="id_3_3_3_2_1_2"></a>
#### [Code](#id_3_3_3_2_1_2)
[Back](#id_3_3_3_2_1)

In [39]:
# merging enhanced_archive_clean and retweet_favorite_counts_clean on tweet_id variable 
# and 'left' type (keep the unmatches from enhanced_archive_clean) 
enhanced_archive_clean = enhanced_archive_clean.merge(retweet_favorite_counts_clean, on='tweet_id', how='left')

<a id="id_3_3_3_2_1_3"></a>
#### [Test](#id_3_3_3_2_1_3)
[Back](#id_3_3_3_2_1)

In [40]:
enhanced_archive_clean.columns.values # Check variable names

array(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id',
       'timestamp', 'source', 'text', 'retweeted_status_id',
       'retweeted_status_user_id', 'retweeted_status_timestamp',
       'expanded_urls', 'rating_numerator', 'rating_denominator', 'name',
       'doggo', 'floofer', 'pupper', 'puppo', 'retweet_count',
       'favorite_count'], dtype=object)

<a id="id_3_3_3_2_2"></a>
### [Metling Dog Stage Columns](#id_3_3_3_2_2)
The 4 "stages" variables (`doggo`, `floofer`, `pupper` and `puppo`) are unnecessary, and can be reduced to one variable named `stage`.<br>
[Define](#id_3_3_3_2_2_1) [Code](#id_3_3_3_2_2_2) [Test](#id_3_3_3_2_2_3)

[Back to Cleaning for Tidiness](#id_3_3_3_2)<br>
[Back to Docummentions](#id_3_2_4_2)

<a id="id_3_3_3_2_2_1"></a>
#### [Define](#id_3_3_3_2_2_1)
Clean `doggo`, `floofer`, `pupper` and `puppo` columns repetition, then melt them into one column named `stage` using [DataFrame.melt](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.melt.html?highlight=melt#pandas.DataFrame.melt).

[Back](#id_3_3_3_2_2)

<a id="id_3_3_3_2_2_2"></a>
#### [Code](#id_3_3_3_2_2_2)
[Back](#id_3_3_3_2_2)

In [41]:
import numpy as np

# get records' indices that have repetition in stage then drop them from enhanced_archive_clean
droped_repeated_ind = repeated_stage.query('repeated > 1').index.values
enhanced_archive_clean.drop(droped_repeated_ind, inplace=True)

# Stage cols names
stage_cols = ['doggo', 'floofer', 'pupper', 'puppo']

# specify the cols to compare with in melf function (all columns except for stage_cols)
id_vars = enhanced_archive_clean.columns.drop(stage_cols)

# melt stage_cols into one col named "stage", then drop the variable col
enhanced_archive_clean = enhanced_archive_clean.melt(id_vars=id_vars, 
                                                         value_vars=stage_cols, 
                                                         value_name='stage').drop(columns=['variable'])

# specify subset columns to get the duplicate columns by (all columns except stage col).
subset_cols = enhanced_archive_clean.columns.drop('stage').values

# sorting values by stage (so we can keep first occurs only later with our dropping a record that have stage value),
# dropping all duplicate records except first occurs, then reset index.
enhanced_archive_clean = enhanced_archive_clean.sort_values(by='stage').drop_duplicates(subset=subset_cols, 
                                                                                        keep='first').reset_index(drop=True)

<a id="id_3_3_3_2_2_3"></a>
#### [Test](#id_3_3_3_2_2_3)
[Back](#id_3_3_3_2_2)

In [42]:
# asserting if stage in the new modified dataframe or not
assert 'stage' in enhanced_archive_clean.columns

# asserting if any of the old stage columns still in the new modified dataframe
assert any(col in stage_cols for col in enhanced_archive_clean.columns) == False

# Value counts
enhanced_archive_clean.stage.value_counts()

pupper     245
doggo       83
puppo       29
floofer      9
Name: stage, dtype: int64

<a id="id_3_3_3_2_3"></a>
### [Melting Predictions Columns](#id_3_3_3_2_3)
Variables `p1` through `p3_dog` in `image_predictions` are unnecessary, and can be reduced to 4 variables.<br>
[Define](#id_3_3_3_2_3_1) [Code](#id_3_3_3_2_3_2) [Test](#id_3_3_3_2_3_3)

[Back to Cleaning for Tidiness](#id_3_3_3_2)<br>
[Back to Docummentions](#id_3_2_4_2)

<a id="id_3_3_3_2_3_1"></a>
#### [Define](#id_3_3_3_2_3_1)
By using [DataFrame.melt](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.melt.html?highlight=melt#pandas.DataFrame.melt), create multi-level melting to add at each level a new variable. to create these new variables:
- **number** as the number of the prediction (1, 2, 3).
- **prediction** as the value of the prediction.
- **conf** as the confidence interval for the prediction
- **reality** as the truth of the prediction.<br><br>


[Back](#id_3_3_3_2_3)

<a id="id_3_3_3_2_3_2"></a>
#### [Code](#id_3_3_3_2_3_2)
[Back](#id_3_3_3_2_3)

In [43]:
# Melting the old predictions dataframe to the new one
df1 = pd.melt(image_predictions_clean, id_vars='tweet_id', 
                                       value_vars=['p1', 'p2', 'p3'], 
                                       var_name='number', 
                                       value_name='prediction')

df2 = df1.join(pd.melt(image_predictions_clean, id_vars='tweet_id', value_vars=['p1_conf', 'p2_conf', 'p3_conf'],  
                        var_name='var_conf',  value_name='conf'), rsuffix='_right').drop(columns=['tweet_id_right', 'var_conf'])

image_predictions_clean = df2.join(pd.melt(image_predictions_clean, id_vars='tweet_id', 
                                           value_vars=['p1_dog', 'p2_dog', 'p3_dog'], 
                                           var_name='var_reality', 
                                           value_name='reality'), rsuffix='_right').drop(columns=['tweet_id_right', 
                                                                                                  'var_reality'])

image_predictions_clean = image_predictions_clean.sort_values(by=['tweet_id', 'number']).reset_index(drop=True)

<a id="id_3_3_3_2_3_3"></a>
#### [Test](#id_3_3_3_2_3_3)
[Back](#id_3_3_3_2_3)

In [44]:
image_predictions_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6225 entries, 0 to 6224
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   tweet_id    6225 non-null   int64  
 1   number      6225 non-null   object 
 2   prediction  6225 non-null   object 
 3   conf        6225 non-null   float64
 4   reality     6225 non-null   bool   
dtypes: bool(1), float64(1), int64(1), object(2)
memory usage: 200.7+ KB


<a id="id_3_3_3_3"></a>
# [3. Cleaning for Quality](#id_3_3_3_3)
Now it's time to clean the dirty tables!!
1. [Enhanced Archive Table](#id_3_3_3_3_1)
2. [Image Predictions Table](#id_3_3_3_3_2)

[Back to Data Cleaning Process](#id_3_3_3)<br>
[Back to Assessment Docummentions](#id_3_2_4)

<a id="id_3_3_3_3_1"></a>
## [1. Enhanced Archive Table](#id_3_3_3_3_1)
1. [Dropping Source and Rating Denominator Columns](#id_3_3_3_3_1_1)
2. [Modifying Name Column](#id_3_3_3_3_1_2)
3. [Modifying Stage Column](#id_3_3_3_3_1_3)
4. [Modifying Expanded Urls Column](#id_3_3_3_3_1_4)
5. [Modifying Enhanced Archive Table Data Types](#id_3_3_3_3_1_5)

[Back to Cleaning for Quality](#id_3_3_3_3)<br>
[Back to Docummentions](#id_3_2_4_1)

In [45]:
enhanced_archive_clean

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     890240255349198849                    NaN                  NaN   
1     758828659922702336                    NaN                  NaN   
2     760521673607086080                    NaN                  NaN   
3     760893934457552897                    NaN                  NaN   
4     763956972077010945           7.638652e+17           15846407.0   
...                  ...                    ...                  ...   
2337  666049248165822465                    NaN                  NaN   
2338  666044226329800704                    NaN                  NaN   
2339  666033412701032449                    NaN                  NaN   
2340  666029285002620928                    NaN                  NaN   
2341  666020888022790149                    NaN                  NaN   

                      timestamp  \
0     2017-07-26 15:59:51 +0000   
1     2016-07-29 00:57:05 +0000   
2     2016-08-02 17:04:31 +0000   
3     2016-08-03 17:43:45 +0000   
4     2016-08-12 04:35:10 +0000   
...                         ...   
2337  2015-11-16 00:24:50 +0000   
2338  2015-11-16 00:04:52 +0000   
2339  2015-11-15 23:21:54 +0000   
2340  2015-11-15 23:05:30 +0000   
2341  2015-11-15 22:32:08 +0000   

                                                                                  source  \
0     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2                        <a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>   
3     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
...                                                                                  ...   
2337  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2338  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2339  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2340  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2341  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                               text  \
0     This is Cassie. She is a college pup. Studying international doggo communication and stick theory. 14/10 so elegant much sophisticate https://t.co/t1bfwz5S2A   
1                                      This doggo is just waiting for someone to be proud of her and her accomplishment. 13/10 legendary af https://t.co/9T2h14yn4Q   
2                                                                Doggo want what doggo cannot have. Temptation strong, dog stronger. 12/10  https://t.co/IqyTF6qik6   
3                               This is Wishes. He has the day off. Daily struggles of being a doggo have finally caught up with him. 11/10 https://t.co/H9YgrUkYwa   
4                                                                                 @TheEllenShow I'm not sure if you know this but that doggo right there is a 12/10   
...                                                                                                                                                             ...   
2337                                       Here we have a 1949 1st generation vulpix. Enjoys sweat tea and Fox News. Cannot be phased. 5/10 https://t.co/4B7cOc1EDq   
2338                      This is a purebred Piers Morgan. Loves to Netflix and chill. Always looks like he forgot to unplug the iron. 6/10 https://t.co/DWnyCjf2mx   
2339                             Here is a very happy pup. Big fan of well-maintained decks. Just look at that tongue. 9/1

<a id="id_3_3_3_3_1_1"></a>
### [1. Dropping Source and Rating Denominator Columns](#id_3_3_3_3_1_1)
`source` variable is irrelevant to our data, and there's no need for it.
<br>
[Define](#id_3_3_3_3_1_1_1) [Code](#id_3_3_3_3_1_1_2) [Test](#id_3_3_3_3_1_1_3)

[Back to Enhanced Archive Table](#id_3_3_3_3_1)<br>
[Back to Docummentions](#id_3_2_4_1)

<a id="id_3_3_3_3_1_1_1"></a>
#### [Define](#id_3_3_3_3_1_1_1)
Use `drop` method to drop `source` and `rating_denominator` variables.

[Back](#id_3_3_3_3_1_1)

<a id="id_3_3_3_3_1_1_2"></a>
#### [Code](#id_3_3_3_3_1_1_2)
[Back](#id_3_3_3_3_1_1)

In [46]:
enhanced_archive_clean.drop(columns=['source', 'rating_denominator'], inplace=True) # Drop unwanted columns

<a id="id_3_3_3_3_1_1_3"></a>
#### [Test](#id_3_3_3_3_1_1_3)
[Back](#id_3_3_3_3_1_1)

In [47]:
assert 'source' not in enhanced_archive_clean.columns
assert 'rating_denominator' not in enhanced_archive_clean.columns

<a id="id_3_3_3_3_1_2"></a>
### [2. Modifying Name Column](#id_3_3_3_3_1_2)
Let's modify name's missing and wrong values<br>
[Define](#id_3_3_3_3_1_2_1) [Code](#id_3_3_3_3_1_2_2) [Test](#id_3_3_3_3_1_2_3)

[Back to Enhanced Archive Table](#id_3_3_3_3_1)<br>
[Back to Docummentions](#id_3_2_4_1)

<a id="id_3_3_3_3_1_2_1"></a>
#### [Define](#id_3_3_3_3_1_2_1)
Use `replace` method to replace missed names values in name column to be Null. Also correct the other wrong names using `lambda` function.

[Back](#id_3_3_3_3_1_2)

<a id="id_3_3_3_3_1_2_2"></a>
#### [Code](#id_3_3_3_3_1_2_2)
[Back](#id_3_3_3_3_1_2)

In [48]:
# Getting the incorrect names that have their original names in the 'text' variable
wrong_fetched_names = enhanced_archive_clean[(enhanced_archive_clean['name'] == 'a') &
                                             (enhanced_archive_clean['text'].str.contains("named"))][['text', 'name']]

# Fetching the right names from 'text' variable
right_fetched_names = wrong_fetched_names.text.str.extract('named.([A-Za-z]+)').rename(columns={0: "name"})

# Getting the right name indices and values
right_names_index = right_fetched_names.index
right_names = right_fetched_names['name']

# Modifying the incorrect names with the correct ones.
enhanced_archive_clean.loc[right_names_index, 'name'] = right_names

In [49]:
# Replcae wrong names with np.nan
enhanced_archive_clean.name = enhanced_archive_clean.name.replace(wrong_names, np.nan)

# Modify some wrong fetched names.
enhanced_archive_clean.name = enhanced_archive_clean.name.apply(lambda x: "O'Malley" if x == "O" 
                                                                                         else('Jo. Jo' if x == 'Jo' 
                                                                                              else(x)))

<a id="id_3_3_3_3_1_2_3"></a>
#### [Test](#id_3_3_3_3_1_2_3)
[Back](#id_3_3_3_3_1_2)

In [50]:
# Check if these names corrected
assert all(name in enhanced_archive_clean.name.values for name in ["O'Malley", 'Jo. Jo']) == True

# Check if there are no lower names left
assert enhanced_archive_clean.loc[enhanced_archive_clean.name.str.islower()==True, 'name'].empty == True

# Check if the incorrect names are corrected
assert enhanced_archive_clean.loc[right_names_index][enhanced_archive_clean.loc[right_names_index]['name'] == 'a'].empty == True

<a id="id_3_3_3_3_1_3"></a>
### [3. Modifying Stage Column](#id_3_3_3_3_1_3)
<br>

[Define](#id_3_3_3_3_1_3_1) [Code](#id_3_3_3_3_1_3_2) [Test](#id_3_3_3_3_1_3_3)

[Back to Enhanced Archive Table](#id_3_3_3_3_1)<br>
[Back to Docummentions](#id_3_2_4_1)

<a id="id_3_3_3_3_1_3_1"></a>
#### [Define](#id_3_3_3_3_1_3_1)
Use astype to change `stage` data type from `object` to `category` and change values case to capital.

[Back](#id_3_3_3_3_1_3)

<a id="id_3_3_3_3_1_3_2"></a>
#### [Code](#id_3_3_3_3_1_3_2)
[Back](#id_3_3_3_3_1_3)

In [51]:
# Modifying data types
enhanced_archive_clean.stage = enhanced_archive_clean.stage.str.capitalize()
enhanced_archive_clean.stage = enhanced_archive_clean.stage.astype('category')

<a id="id_3_3_3_3_1_3_3"></a>
#### [Test](#id_3_3_3_3_1_3_3)
[Back](#id_3_3_3_3_1_3)

In [52]:
enhanced_archive_clean.stage.dtype # Check 'stage' data type

CategoricalDtype(categories=['Doggo', 'Floofer', 'Pupper', 'Puppo'], ordered=False)

<a id="id_3_3_3_3_1_4"></a>
### [4. Modifying Expanded Urls Column](#id_3_3_3_3_1_4)


[Define](#id_3_3_3_3_1_4_1) [Code](#id_3_3_3_3_1_4_2) [Test](#id_3_3_3_3_1_4_3)

[Back to Enhanced Archive Table](#id_3_3_3_3_1)<br>
[Back to Docummentions](#id_3_2_4_1)

<a id="id_3_3_3_3_1_4_1"></a>
#### [Define](#id_3_3_3_3_1_4_1)
Use `lambda` function to iterate over `expanded_urls` variable and `split` on `,` if it was there to fetch one and only one urls for each record.

[Back](#id_3_3_3_3_1_4)

<a id="id_3_3_3_3_1_4_2"></a>
#### [Code](#id_3_3_3_3_1_4_2)
[Back](#id_3_3_3_3_1_4)

In [53]:
# Modifying 'urls' variable
enhanced_archive_clean.expanded_urls = enhanced_archive_clean.expanded_urls.map(lambda x: 
                                                                                    str(x.split(',')[0]) 
                                                                                        if ',' in str(x) 
                                                                                            else(x))

In [54]:
tweet_urls = enhanced_archive_clean.copy()[['tweet_id', 'expanded_urls']] # Create a df with only ids and urls
tweet_urls['url_len'] = tweet_urls['expanded_urls'].map(lambda x: len(str(x))) # Adding url lengths

# url lengths less than 50
tweet_len_le_25 = tweet_urls[(tweet_urls['url_len'] < 25) & (tweet_urls['expanded_urls'].notnull())].index

# Replace values for urls that their length is below 25 (invalid link) 
# except for values with 3 which are (missing values, i.e., NaNs)
enhanced_archive_clean.at[tweet_len_le_25.values[0], 'expanded_urls'] = np.nan

<a id="id_3_3_3_3_1_4_3"></a>
#### [Test](#id_3_3_3_3_1_4_3)
[Back](#id_3_3_3_3_1_4)

In [55]:
enhanced_archive_clean.expanded_urls.map(lambda x: len(str(x))).unique()

array([ 63,  29,   3,  55,  44,  57,  40, 184,  64,  68, 169,  65,  61,
        41,  58,  43,  38,  59,  54,  32,  66,  48,  67,  53,  35,  49,
        50,  51,  69,  45,  56, 173,  86, 123,  52,  34,  37], dtype=int64)

<a id="id_3_3_3_3_1_5"></a>
### [5. Modifying Enhanced Archive Table Data Types](#id_3_3_3_3_1_5)
Change inaccurate `enhanced_archive_clean` variables data types.
<br>
[Define](#id_3_3_3_3_1_5_1) [Code](#id_3_3_3_3_1_5_2) [Test](#id_3_3_3_3_1_5_3)

[Back to Enhanced Archive Table](#id_3_3_3_3_1)<br>
[Back to Docummentions](#id_3_2_4_1)

<a id="id_3_3_3_3_1_5_1"></a>
#### [Define](#id_3_3_3_3_1_5_1)
Use `astype` method to change variables to `object` and `int` data types, and `to_datetime` to datetime data type.

[Back](#id_3_3_3_3_1_5)

<a id="id_3_3_3_3_1_5_2"></a>
#### [Code](#id_3_3_3_3_1_5_2)
[Back](#id_3_3_3_3_1_5)

In [56]:
# Change IDs columns to object data type
enhanced_archive_clean = enhanced_archive_clean.astype({'tweet_id' : 'object',
                               'in_reply_to_status_id' : 'object',
                               'in_reply_to_user_id' : 'object',
                               'retweeted_status_id' : 'object',
                               'retweeted_status_user_id' : 'object'})

# Change retweet_count and favorite_count columns to int after filling NaNs
enhanced_archive_clean.retweet_count = enhanced_archive_clean.retweet_count.fillna(0).astype('int')
enhanced_archive_clean.favorite_count = enhanced_archive_clean.favorite_count.fillna(0).astype('int')

# Change timestamp and retweeted_status_timestamp columns to datetime.
enhanced_archive_clean.timestamp = pd.to_datetime(enhanced_archive_clean.timestamp)
enhanced_archive_clean.retweeted_status_timestamp = pd.to_datetime(enhanced_archive_clean.retweeted_status_timestamp)

<a id="id_3_3_3_3_1_5_3"></a>
#### [Test](#id_3_3_3_3_1_5_3)
[Back](#id_3_3_3_3_1_5)

In [57]:
enhanced_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2342 entries, 0 to 2341
Data columns (total 14 columns):
 #   Column                      Non-Null Count  Dtype              
---  ------                      --------------  -----              
 0   tweet_id                    2342 non-null   object             
 1   in_reply_to_status_id       77 non-null     object             
 2   in_reply_to_user_id         77 non-null     object             
 3   timestamp                   2342 non-null   datetime64[ns, UTC]
 4   text                        2342 non-null   object             
 5   retweeted_status_id         179 non-null    object             
 6   retweeted_status_user_id    179 non-null    object             
 7   retweeted_status_timestamp  179 non-null    datetime64[ns, UTC]
 8   expanded_urls               2282 non-null   object             
 9   rating_numerator            2342 non-null   int64              
 10  name                        1517 non-null   object          

<a id="id_3_3_3_3_2"></a>
## [2. Image Predictions Table](#id_3_3_3_3_2)
1. [Modifying Number Column](#id_3_3_3_3_2_1)
2. [Modifying Prediction Column](#id_3_3_3_3_2_2)
3. [Modifying Image Predictions Table Data Types](#id_3_3_3_3_2_3)

[Back to Cleaning for Quality](#id_3_3_3_3)<br>
[Back to Docummentions](#id_3_2_4_1)

<a id="id_3_3_3_3_2_1"></a>
### [1. Modifying Number Column](#id_3_3_3_3_2_1)

[Define](#id_3_3_3_3_2_1_1) [Code](#id_3_3_3_3_2_1_2) [Test](#id_3_3_3_3_2_1_3)

[Back to Enhanced Archive Table](#id_3_3_3_3_2)<br>
[Back to Docummentions](#id_3_2_4_1)

<a id="id_3_3_3_3_2_1_1"></a>
#### [Define](#id_3_3_3_3_2_1_1)
Use `str.replace` to change `number` variables from `(p1, p2, p3)` to `(1, 2, 3)`.

[Back](#id_3_3_3_3_2_1)

<a id="id_3_3_3_3_2_1_2"></a>
#### [Code](#id_3_3_3_3_2_1_2)
[Back](#id_3_3_3_3_2_1)

In [58]:
image_predictions_clean.number = image_predictions_clean.number.replace(['p1', 'p2', 'p3'], ['1', '2', '3'])

<a id="id_3_3_3_3_2_1_3"></a>
#### [Test](#id_3_3_3_3_2_1_3)
[Back](#id_3_3_3_3_2_1)

In [59]:
image_predictions_clean.number.value_counts()

2    2075
1    2075
3    2075
Name: number, dtype: int64

<a id="id_3_3_3_3_2_2"></a>
### [2. Modifying Prediction Column](#id_3_3_3_3_2_2)

[Define](#id_3_3_3_3_2_2_1) [Code](#id_3_3_3_3_2_2_2) [Test](#id_3_3_3_3_2_2_3)

[Back to Enhanced Archive Table](#id_3_3_3_3_2)<br>
[Back to Docummentions](#id_3_2_4_1)

<a id="id_3_3_3_3_2_2_1"></a>
#### [Define](#id_3_3_3_3_2_2_1)
Use str.replace to modify `prediction` and str.title() to capitalize it.

[Back](#id_3_3_3_3_2_2)

<a id="id_3_3_3_3_2_2_2"></a>
#### [Code](#id_3_3_3_3_2_2_2)
[Back](#id_3_3_3_3_2_2)

In [60]:
image_predictions_clean.prediction = image_predictions_clean.prediction.str.replace('_', ' ').str.title()

<a id="id_3_3_3_3_2_2_3"></a>
#### [Test](#id_3_3_3_3_2_2_3)
[Back](#id_3_3_3_3_2_2)

In [61]:
image_predictions_clean.prediction.unique()[:20]

array(['Welsh Springer Spaniel', 'Collie', 'Shetland Sheepdog', 'Redbone',
       'Miniature Pinscher', 'Rhodesian Ridgeback', 'German Shepherd',
       'Malinois', 'Bloodhound', 'Rottweiler', 'Doberman',
       'Bernese Mountain Dog', 'English Springer',
       'Greater Swiss Mountain Dog', 'Box Turtle', 'Mud Turtle',
       'Terrapin', 'Chow', 'Tibetan Mastiff', 'Fur Coat'], dtype=object)

<a id="id_3_3_3_3_2_3"></a>
### [3. Modifying Image Predictions Table Data Types](#id_3_3_3_3_2_3)

[Define](#id_3_3_3_3_2_3_1) [Code](#id_3_3_3_3_2_3_2) [Test](#id_3_3_3_3_2_3_3)

[Back to Enhanced Archive Table](#id_3_3_3_3_2)<br>
[Back to Docummentions](#id_3_2_4_1)

<a id="id_3_3_3_3_2_3_1"></a>
#### [Define](#id_3_3_3_3_2_3_1)
Use `astype` method to change `tweet_id` data type to `object` and `number` to `category`.

[Back](#id_3_3_3_3_2_3)

<a id="id_3_3_3_3_2_3_2"></a>
#### [Code](#id_3_3_3_3_2_3_2)
[Back](#id_3_3_3_3_2_3)

In [62]:
image_predictions_clean = image_predictions_clean.astype({'tweet_id' : 'object',
                                                          'number' : 'category'})

<a id="id_3_3_3_3_2_3_3"></a>
#### [Test](#id_3_3_3_3_2_3_3)
[Back](#id_3_3_3_3_2_3)

In [63]:
image_predictions_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6225 entries, 0 to 6224
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   tweet_id    6225 non-null   object  
 1   number      6225 non-null   category
 2   prediction  6225 non-null   object  
 3   conf        6225 non-null   float64 
 4   reality     6225 non-null   bool    
dtypes: bool(1), category(1), float64(1), object(2)
memory usage: 158.3+ KB


<a id="id_3_3_4"></a>
> ### [So Far, So Good](#id_3_3_4)
>Now, we have complete wrangling the data set it's time to [Store it](#id_4). Let's cehck [Our Goal](#id_1_3).
>- [X] Gathering
>- [X] Assessing
>- [X] Cleaning
>- [ ] Storing
>- [ ] Analyzing
>- [ ] Visualizing
>- [ ] Reporting
>
>
>[Back to Data Wrangling Process](#id_3)

<a id="id_4"></a>
<h1 style="font-size:30px; color:#3A71AE">Data Storing</h1>

Our data are `complete` and `clean` for now, so, let's store it in a SQL server using [SQLAlchemy](https://www.sqlalchemy.org/) library.

[Back to Top](#top)

In [64]:
from sqlalchemy import create_engine

# specify the database direction
database_dir = os.path.join(folder_data, 'twitter_archive_master.db')

# Create SQLAlchemy Engine and empty we_rate_dogs_tweets database
engine = create_engine('sqlite:///{}'.format(database_dir))

#check if the database file exists and delete it if it do
if os.path.exists(database_dir):
    os.remove(database_dir)

# write each table to a master table in we_rate_dogs_tweets database
enhanced_archive_clean.to_sql('enhanced_archive_master', engine, index=False)
image_predictions_clean.to_sql('image_predictions_master', engine, index=False)

In [65]:
# We can also save the dataset in a csv file

enhanced_archive_clean.to_csv(os.path.join(folder_data, 'enhanced_archive_master.csv'), index=False)
image_predictions_clean.to_csv(os.path.join(folder_data, 'image_predictions_master.csv'), index=False)

<a id="id_4_1"></a>
> ### [So Far, So Good](#id_3_3_4)
>Our data now are stored in two tables `enhanced_archive_master` and `image_predictions_master` in `we_rate_dogs_tweets` database.
>
> In the next notebook [**Analyze and Visualize WeRateDogs**](01_analyze_visualize.ipynb) I will start to analyze the clean data, then visualize what will be found.
>
>- [X] Gathering
>- [X] Assessing
>- [X] Cleaning
>- [x] Storing
>- [ ] Analyzing
>- [ ] Visualizing
>- [ ] Reporting
>
>
>[Back to Data Storing](#id_4)

<img src="src/tbc.png">